src/key_io.cpp

```diff
@@ -62,7 +62,7 @@ class DestinationEncoder
         std::vector<unsigned char> data = {(unsigned char)id.version};
         data.reserve(1 + (id.length * 8 + 4) / 5);
         ConvertBits<8, 5, true>([&](unsigned char c) { data.push_back(c); }, id.program, id.program + id.length);
-        return bech32::Encode(bech32::Encoding::BECH32, m_params.Bech32HRP(), data);
+        return bech32::Encode(bech32::Encoding::BECH32M, m_params.Bech32HRP(), data);
```

Comment:
In commit "Use Bech32m encoding for v1+ segwit addresses" (ea3995aa3cd8f04d7328c95a43c79a5b20fa18a5)

Should AddressDescriptor::GetOutputType should now return BECH32M instead of BECH32 for WitnessUnknown addresses to be consistent with this change?

https://github.com/bitcoin/bitcoin/blob/ea3995aa3cd8f04d7328c95a43c79a5b20fa18a5/src/script/descriptor.cpp#L656

Also, this change seems like the first change that affects user-visible behavior, so it would be great if it added release notes or at least mentioned what the effects are in the commit message. I think the only thing current software can do with V1 addresses is validate them and do watchonly scans for them. So this only affects validate and import RPCs?

Reply:
Very good point.

We can't return `GetOutputType::BECH32M` because no such constant exists. It could be added, but I'd rather not do that in this PR, because I don't think it actually matters at this stage. The purpose of output types is so the user can select what kind of address to request from a wallet. You're right that technically a `addr(...)` descriptor could be imported (but I think that in most cases a ranged descriptor is required, so `addr(...)` won't even work), and it might matter to classify that as bech32m - but this seems like a very unrealistic edge case.

However, I think this will matter once we want to add actual taproot support to the wallet (#21365, and follow-ups that at one point might want to construct such wallet descriptors by default). I suspect we may want an OutputType::BECH32M then, and permit wallets to have separate -addresstype=bech32 and -addresstype=bech32m active descriptors, as one may know that the sender they want to receive from supports bech32 but not bech32m yet.

I've also added release notes.

Reply:
I think `-addresstype=bech32m` makes sense, e.g. `getnewaddress bech32m` as a way to request a taproot p2pk address. However it might be more future proof to use `-addresstype=tap` and  `getnewaddress tap`.

Longer term I suspect it might be better to give wallet descriptors a name and then call `getnewaddress NAME`.

See also #15590 for some earlier discussions around this issue.

Reply:
> However it might be more future proof to use `-addresstype=tap` and `getnewaddress tap`.

I'm not sure. If Bech32m gets adopted as I hope it does, we won't need a new address type for post-taproot things anymore.

Reply:
Since bech32m is (and bech32 was) intended to be used for different segwit versions, I think that we should actually move away from having bech32 or bech32m as the address type as with future segwit versions, it becomes ambiguous as to what type of address you want.

For example, suppose we introduce a segwit v2 that, like segwit v1, will encode a pubkey in the address. Now the wallet would have a descriptor for segwit v1, and another for segwit v2. If I wanted a segwit v1 address, what do I provide as the address type? If I provide `bech32m`, that's ambiguous because both segwit v1 and segwit v2 use `bech32m` addresses. We would have to come up with another name for segwit v2 addresses, and even then, it is still confusing because technically both are bech32m addresses.

With that in mind, I think we need to use a different and more unique naming scheme. Luckily this is pretty easy because segwit has version numbers, and proposals all have unique names. So for segwit v1 addresses, we could call the address type `tap` and just not expose bech32m in it.

I have done a similar thing in HWI where segwit v0 is `wit` rather than `bech32`, and segwit v1 will be `tap`.

Looking at it now, I do also think that having `bech32` as the address type for segwit v0 was not the right thing to do. It really should have just be `wit` or something like that that makes it unique to just segwit v0. Since bech32 is now segwit v0 only, it's fine, but in the future, we should not do the same thing.

Reply:
> Now the wallet would have a descriptor for segwit v1, and another for segwit v2. If I wanted a segwit v1 address, what do I provide as the address type?

I think that's the crucial point: why would you want that? If every sender supports both equally, you'd always produce whatever you have a descriptor for that is compatible with that.

If there is a reason for being more flexible, no objection to just using names or something. But then also, would you support having multiple distinct active descriptors that produce the same address type within the same wallet? I think that's the same problem.

Reply:
> I think that's the crucial point: why would you want that? If every sender supports both equally, you'd always produce whatever you have a descriptor for that is compatible with that.

In the ideal world, sure. But as we have seen with segwit, that's not guaranteed. Even though senders were supposed to allow any segwit version with bech32, they didn't. I'm not sure that everyone will understand to do that with bech32m.

The address type option is not just for senders. It's also for receivers and what they are comfortable with receiving. I could conceive of a situation where someone sets up a new wallet but wants to use v1 instead of v2 because, e.g., v2 introduces some new signature algo that has assumptions they aren't comfortable with using. What would they have to do to use v1?

Reply:
> In the ideal world, sure. But as we have seen with segwit, that's not guaranteed. Even though senders were supposed to allow any segwit version with bech32, they didn't. I'm not sure that everyone will understand to do that with bech32m.

Sure, that's why I suggest we may want to add an OutputType::BECH32M - precisely because Bech32 and Bech32m may differ for senders. But p2sh vs. p2sh-p2wpk shouldn't be distinguished, for example.

> I could conceive of a situation where someone sets up a new wallet but wants to use v1 instead of v2 because, e.g., v2 introduces some new signature algo that has assumptions they aren't comfortable with using. What would they have to do to use v1?

That's fair, and was a good argument for segwit vs non-segwit in pre-descriptor wallets. But in descriptor wallets if you don't want witness v9 addresses because you're not comfortable with them, simply don't import/create such a descriptor.

Reply:
> Sure, that's why I suggest we may want to add an OutputType::BECH32M - precisely because Bech32 and Bech32m may differ for senders.

Yes, I agree with doing that.

> But in descriptor wallets if you don't want witness v9 addresses because you're not comfortable with them, simply don't import/create such a descriptor.

That's reasonable, but I think it's too onerous on users to have to figure out which and how to create the descriptor that they want. I think it's reasonable that user would know about different address types and, at a high level, the consequences of using each type, but also not know (or be afraid of) the descriptor syntax, which derivation paths to use, safely handling xprvs for import, etc. I suppose we could add more options to the wallet creation stuff, but that still means the user has to deal with descriptors directly once they feel comfortable enough to upgrade, or they have to make a wholly new wallet.

If instead the wallet generated v1 and v2 descriptors, then the user could switch to the new address type by choosing the option and it's very simple and easy to do so.

Reply:
I'm going to mark this as resolved; we've discussed this in today's wallet meeting. It's not exactly clear how to proceed, but in any case, I don't think this is something for this PR. Please respond if you don't agree.

Reply:
Discussion: http://www.erisian.com.au/bitcoin-core-dev/log-2021-03-12.html#l-286